rascal-textmate-core/src/main/rascal/lang/oniguruma/Conversion.rsc

```diff
@@ -0,0 +1,196 @@
+    if (!isEmpty(exceptConditions)) {
+        if (/\choice(symbol, alternatives) := g) {
+
+            bool keep(prod(def, _, _))
+                = \label(l, _) := def
+                ? \except(l) notin exceptConditions
+                : true;
+            
+            re = infix("|", toRegExps(g, {a | a <- alternatives, keep(a)}));
+        }
+    }
+
+    // Convert prefix conditions (depends on previous conversions)
+    if (!isEmpty(prefixConditions)) {
+        re = infix("", toRegExps(g, prefixConditions) + [re]);
+    }
+
+    // Convert suffix conditions (depends on previous conversions)
+    if (!isEmpty(suffixConditions)) {
+        re = infix("", [re] + toRegExps(g, suffixConditions));
+    }
+
+    // Convert delete conditions (depends on previous conversions)
+    if (!isEmpty(deleteConditions)) {
+        RegExp delete = infix("|", [toRegExp(g, s) | \delete(s) <- deleteConditions]);
+            
+        // TODO: Explain this complicated conversion...
+        str string = "(?=(?\<head\><re.string>)(?\<tail\>.*)$)(?!(?:<delete.string>)\\k\<tail\>$)\\k\<head\>";
+        list[str] categories = ["", *re.categories, "", *delete.categories];
+        re = regExp(string, categories);
+    }
+
+    return re;
+}
+
+default RegExp toRegExp(Grammar _, Symbol s) {
+    throw "Unsupported symbol <s>";
+}
+
+@synopsis{
+    Converts a condition to a regular expression.
+}
+
+RegExp toRegExp(Grammar g, \follow(symbol))
+    = prefix("(?=", suffix(")", toRegExp(g, symbol)));
+RegExp toRegExp(Grammar g, \not-follow(symbol))
+    = prefix("(?!", suffix(")", toRegExp(g, symbol)));
+RegExp toRegExp(Grammar g, \precede(symbol))
+    = prefix("(?\<=", suffix(")", toRegExp(g, symbol)));
+RegExp toRegExp(Grammar g, \not-precede(symbol))
+    = prefix("(?\<!", suffix(")", toRegExp(g, symbol)));
+RegExp toRegExp(Grammar g, \delete(symbol)) {
+    throw "Presumably unreachable..."; } // Covered by `toRegExp(Grammar g, \conditional(symbol, conditions))`
+RegExp toRegExp(Grammar _, \at-column(n))
+    = regExp("(?\<=$<right("", n, ".")>)", []);
+RegExp toRegExp(Grammar _, \begin-of-line())
+    = regExp("(?:^)", []);
+RegExp toRegExp(Grammar _, \end-of-line())
+    = regExp("(?:$)", []);
+RegExp toRegExp(Grammar _, \except(_)) {
+    throw "Presumably unreachable..."; } // Covered by `toRegExp(Grammar g, \conditional(symbol, conditions))`
+
+default RegExp toRegExp(Grammar _, Condition c) {
+    throw "Unsupported condition <c>";
+}
+
+@synopsis{
+    Converts a character range to a regular expression.
+}
+
+RegExp toRegExp(Grammar _, range(begin, end))
+    = regExp("[<encode(begin)>-<encode(end)>]", []);
+
+@synopsis{
+    Encodes a (list of) char(s) to a (list of) code unit(s).
+}
+
+str encode(list[int] chars, bool withBounds = false)
+    = withBounds
+    ? "\\b<encode(chars, withBounds = false)>\\b"
+    : intercalate("", [encode(i) | i <- chars]);
+
+str encode(int char) {
+    if (char <= 0xFFFF) {
```

Comment:
The threshold is : `0x7FFF`, as the highest bits are used to signal the surrogate pair.